### PR TITLE
Add content upload/download using AzureBlobStorage

### PR DESCRIPTION
**Pull request by PublicFedorenko**
Usage example:

image
```
public ReservationService(ICancellationTokenProvider cancellationTokenProvider, IBlobStorageService blobStorageService) {
    // upload Reservations
    cancellationTokenProvider.CancellationToken.Register(() => {
        if (Reservations is not null) {
            blobStorageService.Upload(
                content: Reservations, 
                blobName: nameof(Reservation), 
                overwrite: true);
        }
    });

    // download Reservations
    var reservations = Enumerable.Empty<Reservation>();
    if (blobStorageService.BlobExists(blobName: nameof(Reservation)).Value) {
        reservations = blobStorageService
            .DownloadContent(blobName: nameof(Reservation)).Value.Content
            .ToObjectFromJson<IEnumerable<Reservation>>();
    }
    Reservations = new ConcurrentBag<Reservation>(reservations);
}
```

### CODE REVIEW

File: FaynaTown.Tennis.Planner.Bot.Core/Services/IBlobStorageService.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿using System.Threading.Tasks;
+using Azure;
+using Azure.Storage.Blobs.Models;
+
+namespace FaynaTown.Tennis.Planner.Bot.Core.Service;
```
Comment (p1v0zavr) [inline FaynaTown.Tennis.Planner.Bot.Core/Services/IBlobStorageService.cs:5]:
```suggestion
namespace FaynaTown.Tennis.Planner.Bot.Core.Services;
```

File: FaynaTown.Tennis.Planner.Bot.Core/Services/IBlobStorageService.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+public interface IBlobStorageService {
+    Response<BlobContentInfo> Upload(object content, string blobName, bool overwrite = false);
+    Response<BlobDownloadResult> DownloadContent(string blobName);
+    Response<bool> BlobExists(string blobName);
```
Comment (p1v0zavr) [inline FaynaTown.Tennis.Planner.Bot.Core/Services/IBlobStorageService.cs:10]: Do we need all of these methods? Can't we just leave async ones?

File: FaynaTown.Tennis.Planner.Bot.Core/Providers/IFileExtensionProvider.cs [hunk 3]
```diff
@@ -0,0 +1,4 @@
+﻿namespace FaynaTown.Tennis.Planner.Bot.Core.Providers;
+public interface IFileExtensionProvider {
```
Comment (p1v0zavr) [inline FaynaTown.Tennis.Planner.Bot.Core/Providers/IFileExtensionProvider.cs:2]: Little format refactoring suggestion

```suggestion
﻿namespace FaynaTown.Tennis.Planner.Bot.Core.Providers;
﻿
public interface IFileExtensionProvider {
```

File: FaynaTown.Tennis.Planner.Bot.Core/Providers/JsonExtensionProvider.cs [hunk 4]
```diff
@@ -0,0 +1,6 @@
+﻿using System.IO;
+
+namespace FaynaTown.Tennis.Planner.Bot.Core.Providers;
+internal class JsonExtensionProvider : IFileExtensionProvider {
```
Comment (p1v0zavr) [inline FaynaTown.Tennis.Planner.Bot.Core/Providers/JsonExtensionProvider.cs:4]: Little format refactoring suggestion

```suggestion
namespace FaynaTown.Tennis.Planner.Bot.Core.Providers;

internal class JsonExtensionProvider : IFileExtensionProvider {
```

File: FaynaTown.Tennis.Planner.Bot.Core/Providers/JsonExtensionProvider.cs [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+    public string ChangeExtension(string path) => Path.ChangeExtension(path, ".json");
```
Comment (p1v0zavr) [inline FaynaTown.Tennis.Planner.Bot.Core/Providers/JsonExtensionProvider.cs:5]: Little format refactoring suggestion: if we use new line approach for lambda methods everywhere, we use it everywhere :)

```suggestion
    public string ChangeExtension(string path) =>
        Path.ChangeExtension(path, ".json");
```

File: FaynaTown.Tennis.Planner.Bot.Core/Services/BlobStorageService.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    public async Task<Response<bool>> BlobExistsAsync(string blobName) =>
+        await GetBlobClient(blobName).ExistsAsync();
+
+    private BlobClient GetBlobClient(string blobName) {
+        return new BlobClient(
+            connectionString: BlobStorageConnectionString,
+            blobContainerName: ContainerName,
+            blobName: FileExtensionProvider.ChangeExtension(blobName));
+    }
+}
```
Comment (p1v0zavr) [inline FaynaTown.Tennis.Planner.Bot.Core/Services/BlobStorageService.cs:57]: Why do we create new instance for every existance verification/upload/download? Does it affects the cost somehow? Is it time consuming operation for opening connection every time? Maybe it worth it to introduce another level of abstraction where we have one BlobClient instance for every needed blob?

File: FaynaTown.Tennis.Planner.Bot.Core/Services/BlobStorageService.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    public Response<BlobContentInfo> Upload(object content, string blobName, bool overwrite = false) {
+        return GetBlobClient(blobName).Upload(
+            content: BinaryData.FromObjectAsJson(content),
+            overwrite: overwrite);
+    }
+
+    public Response<BlobDownloadResult> DownloadContent(string blobName) {
+        return GetBlobClient(blobName).DownloadContent();
+    }
+    public Response<bool> BlobExists(string blobName) =>
```
Comment (p1v0zavr) [inline FaynaTown.Tennis.Planner.Bot.Core/Services/BlobStorageService.cs:35]:
```suggestion
    }
    
    public Response<bool> BlobExists(string blobName) =>
```

File: FaynaTown.Tennis.Planner.Bot.Core/Configuration/CoreConfigurationModule.cs [hunk 8]
```diff
@@ -3,6 +3,7 @@
 using FaynaTown.Tennis.Planner.Bot.Core.Handlers;
 using FaynaTown.Tennis.Planner.Bot.Core.Helpers;
 using FaynaTown.Tennis.Planner.Bot.Core.Providers;
+using FaynaTown.Tennis.Planner.Bot.Core.Service;
```
[the remark this answers is not in the document]
Comment (PublicFedorenko) [reply FaynaTown.Tennis.Planner.Bot.Core/Configuration/CoreConfigurationModule.cs:6]: Done

File: FaynaTown.Tennis.Planner.Bot.Core/FaynaTown.Tennis.Planner.Bot.Core.csproj [hunk 9]
```diff
@@ -11,6 +11,7 @@
 	<ItemGroup>
 		<PackageReference Include="Ardalis.GuardClauses" Version="4.0.1" />
 		<PackageReference Include="Autofac.Extensions.DependencyInjection" Version="8.0.0" />
+		<PackageReference Include="Azure.Storage.Blobs" Version="12.17.0" />
```
[the remark this answers is not in the document]
Comment (PublicFedorenko) [reply FaynaTown.Tennis.Planner.Bot.Core/FaynaTown.Tennis.Planner.Bot.Core.csproj:14]: Done

File: FaynaTown.Tennis.Planner.Bot.Core/Services/IBlobStorageService.cs (hunk 1 above)
Comment (PublicFedorenko) [reply FaynaTown.Tennis.Planner.Bot.Core/Services/IBlobStorageService.cs:5]: Done

File: FaynaTown.Tennis.Planner.Bot.Core/Providers/IFileExtensionProvider.cs (hunk 3 above)
Comment (PublicFedorenko) [reply FaynaTown.Tennis.Planner.Bot.Core/Providers/IFileExtensionProvider.cs:2]: Done

File: FaynaTown.Tennis.Planner.Bot.Core/Providers/JsonExtensionProvider.cs (hunk 4 above)
Comment (PublicFedorenko) [reply FaynaTown.Tennis.Planner.Bot.Core/Providers/JsonExtensionProvider.cs:4]: Done

File: FaynaTown.Tennis.Planner.Bot.Core/Providers/JsonExtensionProvider.cs (hunk 5 above)
Comment (PublicFedorenko) [reply FaynaTown.Tennis.Planner.Bot.Core/Providers/JsonExtensionProvider.cs:5]: Done

File: FaynaTown.Tennis.Planner.Bot.Core/Services/BlobStorageService.cs (hunk 6 above)
Comment (PublicFedorenko) [reply FaynaTown.Tennis.Planner.Bot.Core/Services/BlobStorageService.cs:57]: **- Why do we create new instance for every existance verification/upload/download?** 
To work with a blob we need a `BlobClient`, which corresponds to it. The only way to configure `BlobClient` is through its constructor. `blobName` property of the existing `BlobClient` is _read-only_. 
Thus, we have to create a new client for each blob and cannot reuse the same `BlobClient`.
**- Does it affects the cost somehow?** 
 No
**- Is it time consuming operation for opening connection every time?** 
 Not sure, but I just found out that MS [recommends](https://learn.microsoft.com/en-us/azure/storage/blobs/storage-blob-client-management?tabs=dotnet#manage-client-objects) storing clients somehow.
**- Maybe it worth it to introduce another level of abstraction where we have one BlobClient instance for every needed blob?**
Yes, some kind of `Factory`. I'll see what we can do.

File: FaynaTown.Tennis.Planner.Bot.Core/Services/BlobStorageService.cs (hunk 7 above)
Comment (PublicFedorenko) [reply FaynaTown.Tennis.Planner.Bot.Core/Services/BlobStorageService.cs:35]: Done

File: FaynaTown.Tennis.Planner.Bot.Azure/local.settings.json [hunk 10]
```diff
@@ -4,6 +4,8 @@
     "AzureWebJobsStorage": "UseDevelopmentStorage=true",
     "FUNCTIONS_WORKER_RUNTIME": "dotnet-isolated",
     "AzureKeyVaultUri": "https://faynatown-keyvault-dev.vault.azure.net", // TODO: change the key vault name
+    "AzureBlobStorageConnectionString": "", // Insert your value for test purposes only
+    "AzureBlobStorageContainerName": "fayna-town-planner-storage",
```
[the remark this answers is not in the document]
Comment (PublicFedorenko) [reply FaynaTown.Tennis.Planner.Bot.Azure/local.settings.json:8]: Todo is a no-problem.
Can you clarify why we need to swap the lines order?

[the remark this answers is not in the document]
Comment (PublicFedorenko) [reply FaynaTown.Tennis.Planner.Bot.Azure/local.settings.json:8]: Added the "TODO:"

File: FaynaTown.Tennis.Planner.Bot.Core/Services/IBlobStorageService.cs (hunk 2 above)
Comment (PublicFedorenko) [reply FaynaTown.Tennis.Planner.Bot.Core/Services/IBlobStorageService.cs:10]: It's better to have additional synchronous methods than call `async` methods with `.GetAwaiter().GetResult()`.

File: FaynaTown.Tennis.Planner.Bot.Core/Services/BlobStorageService.cs (hunk 6 above)
Comment (PublicFedorenko) [reply FaynaTown.Tennis.Planner.Bot.Core/Services/BlobStorageService.cs]: I added `BlobClientFactoryWithCache` to handle BlobClient creation.

**Pull request merged**